### DISCUSSION TITLE
Collecting metrics from plausible (with Prometheus)

### DISCUSSION THREAD

**Discussion by digital88** [plausible/analytics#3638] [Self-Hosted Support] [answered]
Hello.

Does plausible have a way to grab internal metrics with tools like Prometheus? I see some metrics generation code here https://github.com/plausible/analytics/blob/master/lib/plausible/telemetry/plausible_metrics.ex

I tried to make a GET request to /metrics endpoint (usually the "default" way of getting metrics from services) but got "Service Unavailable". Is there a way to get metrics from self-hosted instance?

**Comment by ruslandoga** [comment, marked as answer]
👋 @digital88 

I think you'd need to enable [PromEx](https://github.com/akoutmos/prom_ex) by setting [`PROMEX_DISABLED=false`](https://github.com/plausible/analytics/blob/9e783ebb418f3a212ee9b1182b906133803384b5/config/runtime.exs#L672) in your env vars.

**Reply by ruslandoga** [reply]
Related:

- you can cross reference [`Plausible.PromEx`](https://github.com/plausible/analytics/blob/9e783ebb418f3a212ee9b1182b906133803384b5/lib/plausible/prom_ex.ex) against PromEx docs to see what's available
- [`PromEx.Plug`](https://github.com/plausible/analytics/blob/9e783ebb418f3a212ee9b1182b906133803384b5/lib/plausible_web/endpoint.ex#L56) doesn't have `:path` set, so it'll default to [`/metrics`](https://hexdocs.pm/prom_ex/PromEx.Plug.html) as expected

**Reply by digital88** [reply]
Thanks, I'll try! Is there any way to "hot-reload" PROMEX_DISABLE env var in running container? Maybe with `/bin/palusible remote` ?

**Reply by ruslandoga** [reply]
It's complicated :) I think it's easier to recreate the container.

**Reply by ruslandoga** [reply]
Ah, no. There is an easy way but I don't know about all the ⚠️ implications ⚠️ of restarting the application this way.

I've enabled https://who.edify.space/metrics within the running container by setting the env var and then calling [`:init.restart`](https://www.erlang.org/doc/man/init#restart-0)

```console
$ cd hosting
$ docker compose exec plausible bin/remote
```
```elixir
iex(plausible@26dacc616641)1> System.put_env("PROMEX_DISABLED", "false")
:ok
iex(plausible@26dacc616641)2> :init.restart
:ok
# *** ERROR: Shell process terminated! (^G to start new job) ***
```
That error is expected since `IEx` (interactive Elixir) is also restarted.
Press <kbd>ctrl+c</kbd> twice to exit the "dead" IEx console and check if `/metrics` is enabled.

---

```console
$ docker ps --filter name=hosting-plausible-1
CONTAINER ID   IMAGE                        COMMAND                  CREATED       STATUS       PORTS                      NAMES
26dacc616641   plausible/analytics:v2.0.0   "/entrypoint.sh sh -…"   2 weeks ago   Up 2 weeks   127.0.0.1:8000->8000/tcp   hosting-plausible-1
```
Docker didn't notice the restart since the application was restarted within the same Erlang BEAM process.

**Reply by digital88** [reply]
I just tried this on staging, with container restart. Yes, it works. I will probably just restart my prod container too. Thanks again for help!

Btw, it's such a great feature, why its not documented here https://plausible.io/docs/self-hosting-configuration ? :)
